### DISCUSSION TITLE
setPWM  latency

### DISCUSSION THREAD

**Discussion by jmdodd95682** [khoih-prog/RP2040_PWM#19] [General]
I noticed that executing the setPWM command has approximately a 32us latency.  Which I guess would normally be fine, but if you're trying to use it to generate audio, this is too slow.  You need to update the duty cycle of the PMW out faster than that.  Or at least I would want it to be faster for my purposes.  Is this expected?

I thought, at first, it was somehow related to the PWM frequency, but regardless of duty cycle and frequency the latency is about 32us. It is a very tight loop.  The only other things in the loop take about 1us.  The output pin in the code below is toggling at about 29KHz.  When I comment out the setPWM it toggles at about 490KHz.

void loop()
{
  digitalWrite(digPin,dig_output);
  dig_output^=0x01;
  
  data=testpattern[testdata_count];
  testdata_count=(testdata_count+1)%624;

  PWM_Instance->setPWM_Int(pinToUse, frequency, data);
}

**Comment by khoih-prog** [comment]
I suggest you to use the `setPWM_manual()` function to be more efficient as well as rewrite your inefficient code ( e.g. repetitive digitalWrite() iin loop() )

https://github.com/khoih-prog/RP2040_PWM/blob/353f9e340e7bf43fbd47801180a37d9a4f3208fe/src/RP2040_PWM.h#L279-L280

Check these examples to see how you can use and adapt if OK

1. [**PWM_manual**](https://github.com/khoih-prog/RP2040_PWM/tree/main/examples/PWM_manual)
2. [**PWM_Waveform_Fast**](https://github.com/khoih-prog/RP2040_PWM/tree/main/examples/PWM_Waveform_Fast)

If still not fast enough, you have to use `PIO` directly

**Reply by jmdodd95682** [reply]
Thanks!  Yes, the digitalWrites were only in there to measure the latency of the setPWM.  The real loop is just a pop from a buffer and a setPWM.

I will try the setPWM_manual.

**Comment by jmdodd95682** [comment]
I tried it out.  It is much faster.  It is approximately 2us instead of 32us .  Thanks again for the help.

**Reply by khoih-prog** [reply]
Good news. Hopefully it's OK for your use case now.

If I have time, I'll **optimize** the library a little bit more to drop all the unnecessary code.

Will let you know then.

**Comment by khoih-prog** [comment]
Hi @jmdodd95682 

Just modify the code and I'm able to reduce nearly 50%. Now the new `setPWM_manual_Fast()` consumes 1597ns, compare to 2889ns of `setPWM_manual()`

I'll publish the new release ASAP for you to test

---

### Releases v1.6.0

1. Optimize speed with new `setPWM_manual_Fast` function to improve almost 50% compared to `setPWM_manual`. Check 
2. Add example [**PWM_SpeedTest**](https://github.com/khoih-prog/RP2040_PWM/tree/main/examples/PWM_SpeedTest) to demo the better speed of new `setPWM_manual_Fast` function
3. Modify examples [**PWM_manual**](https://github.com/khoih-prog/RP2040_PWM/tree/main/examples/PWM_manual) to use new `setPWM_manual_Fast` function

**Reply by jmdodd95682** [reply]
Wow!  Very nice.  Thank you.  I will check it out.

**Comment by khoih-prog** [comment]
Hi @jmdodd95682 

The new [**RP2040_PWM** v1.6.0](https://github.com/khoih-prog/RP2040_PWM/releases/tag/v1.6.0) has just been published. Your contribution is noted in [**Contributions and Thanks**](https://github.com/khoih-prog/RP2040_PWM#contributions-and-thanks)

Have a  look at the new [**PWM_SpeedTest**](https://github.com/khoih-prog/RP2040_PWM/tree/main/examples/PWM_SpeedTest) example

Please test and don't hesitate to report bug / issue or request enhancement. Even a simple discussion can lead to new and better library features.

Best Regards,
